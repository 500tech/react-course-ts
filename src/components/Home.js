import React from 'react';

export function Home() {
  return (
    <div>
      <h1>This is HOME</h1>
    </div>
  );
}
