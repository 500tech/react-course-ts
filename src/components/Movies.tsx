import * as React from 'react';
import Button from './common/Button';
import Input from './common/Input';
import Movie from './Movie';

interface ComponentProps {
  data: Array<{
    label: string,
    rating: number,
    id: number
  }>,
  addMovie: () => void,
  clearAll: () => void,
  removeMovie: (id) => void,
  selectMovie: (id) => void,
  updateValue: (val) => void,
  inputVal: string
}

const Movies = (props: ComponentProps) => (
  <div className="movies">
    <div className="header">my favorite movies</div>

    <Input onChange={props.updateValue} value={props.inputVal} />

    <Button onClick={props.addMovie}>add movie</Button>

    {
      props.data.length ? (
        <ul>
          {
            props.data.map(movie => (
              <Movie
                key={movie.id}
                movie={movie}
                selectMovie={props.selectMovie}
                removeMovie={props.removeMovie}
              />
            ))
          }
        </ul>
      ) : (
        <div className="empty">
          no movies
        </div>
      )
    }

    <div className="filters">
      <div className="clear" onClick={props.clearAll}>clear all</div>
    </div>
  </div>
);

export default Movies;
