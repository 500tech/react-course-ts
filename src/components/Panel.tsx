import * as React from 'react';
import Rating from './common/Rating';
// import {fetchMovieDetails} from "../queries/movies";
// import api from '../utils/api';

interface ComponentProps {
  movie: {
    label: string,
    rating: number,
    id: number,
    description: string
  },
  updateRating: (id, val) => void,
  clearSelected: () => void
}

const Panel = (props: ComponentProps) => {
  if (!props.movie) {
    return null;
  }

  return (
    <div className="panel">
      <div className="panel-title">{props.movie.label}</div>
      <div className="panel-description">
        {props.movie.description}
      </div>

      <Rating
        amount={props.movie.rating}
        updateRating={val => props.updateRating(props.movie.id, val)} />

      <div
        className="panel-close"
        onClick={props.clearSelected}>
        close panel
      </div>
    </div>
  );
};

export default Panel;
