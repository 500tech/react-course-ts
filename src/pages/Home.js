import React from 'react';

export function Home() {
  return <p>This is the home page</p>;
}
