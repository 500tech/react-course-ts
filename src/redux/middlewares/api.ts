const api = store => next => action => {
  console.log('here');
  fetch('https://jsonplaceholder.typicode.com/users')
    .then(response => {
      console.log(response);
      response.json();
    })
    .then()
    .catch(error => console.log(`Network error:`, error));

  return next(action);
};

export default api;
