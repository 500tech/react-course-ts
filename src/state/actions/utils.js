export const actionCreator = (type, metaFactory) => payload => ({
  type,
  payload,
  meta: metaFactory && metaFactory(payload),
});
