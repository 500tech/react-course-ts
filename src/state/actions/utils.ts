import { createAction, PayloadAction } from "redux-starter-kit";
import { AxiosRequestConfig } from "axios";

export type ApiLabel = AxiosRequestConfig & {
  label?: string;
  onSuccess?: string;
  onFailure?: string;
};

export interface APIAction<Payload = any> extends PayloadAction<Payload> {
  meta: {
    api: ApiLabel | ((payload: Payload) => ApiLabel);
  };
}

export function createApiAction<Payload = undefined>(
  actionType: string,
  apiLabel: APIAction["meta"]["api"]
) {
  const actionCreator = createAction<Payload>(actionType) as (
    payload: Payload
  ) => PayloadAction<Payload>;
  const apiActionCreator = (payload: Payload) => {
    const action = actionCreator(payload);
    return Object.assign(action, {
      meta: {
        api: apiLabel
      }
    }) as APIAction<Payload>;
  };
  apiActionCreator.toString = () => actionType;
  return apiActionCreator;
}
