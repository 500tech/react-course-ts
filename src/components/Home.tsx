import {find, findIndex} from 'lodash';
import * as React from 'react';

import { executeQuery } from '../utils/api';
import Movies from './Movies';
import Panel from './Panel';
import {fetchMovies} from "../queries/movies";

interface ComponentState {
  data: Array<{
    label: string,
    rating: number,
    id: number
  }>,
  inputVal: string,
  selectedId: null | number
}

interface ComponentProps {
  match: any
}

class Home extends React.Component<ComponentProps, ComponentState> {
  state = {
    data: [],
    inputVal: '',
    selectedId: null
  };

  componentDidMount() {
    executeQuery(fetchMovies)
        .then((payload) => {
          this.setState({ data: payload.data.allMovies});
        });

    if (typeof this.props.match.params.id === 'string') {
      this.setState({ selectedId: parseInt(this.props.match.params.id, 10) });
    }
  }

  addMovie = () => {
    const { inputVal } = this.state;

    this.setState({
      data: this.state.data.concat({
        id: this.state.data.length,
        label: inputVal || `new movie - ${this.state.data.length}`,
        rating: 1
      })
    });
  };

  updateRating = (movieId, val) => {
    const newData = [...this.state.data];
    const movieIndex = findIndex(newData, { id: movieId });

    if (movieIndex > -1) {
      newData[movieIndex] = Object.assign({}, newData[movieIndex], {
        rating: val
      });
    }

    this.setState({ data: newData });
  };

  removeMovie = (id) => {
    this.setState({
      data: this.state.data.filter(movie => movie.id !== id)
    });
  };

  clearAll = () => {
    this.setState({
      data: [],
      selectedId: null
    });
  };

  clearSelected = () => {
    this.setState({ selectedId: null });
  };

  updateValue = (e) => {
    const inputVal = e.target.value;

    this.setState({
      inputVal
    });
  };

  render() {
    const { data, selectedId, inputVal } = this.state;
    const activeMovie = find(this.state.data, { id: selectedId });

    return (
      <div className="app">
        <Movies
          data={data}
          inputVal={inputVal}
          addMovie={this.addMovie}
          clearAll={this.clearAll}
          removeMovie={this.removeMovie}
          updateValue={this.updateValue}
          selectMovie={id => this.setState({ selectedId: id })} />

        <Panel
          movie={activeMovie}
          updateRating={this.updateRating}
          clearSelected={this.clearSelected} />
      </div>
    );
  }
}

export default Home;
