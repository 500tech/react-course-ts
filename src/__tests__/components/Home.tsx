import * as React from 'react';
import { shallow } from 'enzyme';
import Home from "../../components/Home";
import {executeQuery} from "../../utils/api";
import Movies from "../../components/Movies";
import Mock = jest.Mock;
import Panel from "../../components/Panel";

/*
  Here we mock the api module, so out tests do not go to our server.
  We generate a mock function with jest and instruct it to return a resolved promise
  with an empty data set by default.

  We will override this implementation in every relevant test
 */
jest.mock('../../utils/api', () => {
  return {
    executeQuery: jest.fn().mockResolvedValue({ data: { allMovies: [] }})
  }
});

/*
  When we test code that runs promises we don't have access to, we don't know how to tell jest when
  the promise was completed and jest can continue the test.
  In order to overcome this, we create a utility function that generates a new promise and resolves it
  on the next processing frame, also known as 'next tick'.

  What happens is that initially we had a call stack with a promise in it. By running a new promise that resolves
  after the next tick, we can know for sure that all previous promises were resolved and we can continue our test.
 */
const scheduler = typeof setImmediate === 'function' ? setImmediate : setTimeout;

function flushPromises() {
  return new Promise(function(resolve) {
    scheduler(resolve);
  });
}

const setup = ({ id = null } = {}) => ({
  match: {
    params: {
      id
    }
  }
});

/*
  A utility function that allows us to change the value of our api result.
  There are more functions like these below that help us return the correct response
  for each api type.
 */
function setupFetchMovies(includeExtraData: boolean) {
  const extraData = {
    description: 'dummy description', rating: 2
  };
  const payload: any = {
    data: {
      allMovies: [
        {id: 0, label: 'movie title'}
      ]
    }
  };

  if (includeExtraData) {
    payload.data.allMovies[0] = { ...payload.data.allMovies[0], ...extraData };
  }

  const executeQueryMock = executeQuery as Mock;
  executeQueryMock.mockResolvedValue(payload);

  return payload;
}

function setupMovieExtraDetailsFetch() {
  const payload = {
    data: {
      Movie: {
        description: 'dummy description',
        rating: 2
      }
    }
  };
  const executeQueryMock = executeQuery as Mock;
  executeQueryMock.mockResolvedValue(payload);

  return payload;
}

function setupCreateMovie() {
  const payload = {
    data: {
      createMovie: {
        id: '1',
        label: 'movie title',
        rating: 1,
        description: 'details about movie'
      }
    }
  };

  (executeQuery as Mock).mockResolvedValue(payload);

  return payload;
}

describe('<Home/>', () => {

  beforeEach(() => {
    (executeQuery as Mock).mockClear();
  });

  test('should render', () => {
    const props = setup();
    const component = shallow(
        <Home {...props} />
    );

    expect(component).toMatchSnapshot();
  });

  test('should fetch movies', async () => {
    const payload = setupFetchMovies(false);
    const props = setup();

    const component = shallow(
        <Home {...props} />
    );

    await flushPromises();
    component.update();

    expect(component.find(Movies)).toHaveProp('data', payload.data.allMovies);
  });

  test('should select a movie and fetch description and rating if missing', async () => {
    const props = setup();
    const payload = setupFetchMovies(false);

    const component = shallow(
        <Home {...props} />
    );

    // wait for all promises to be done before we continue with our test
    await flushPromises();
    const selectMovie = component.find(Movies).prop('selectMovie');

    // clear the mock calls until this point so we can assert the api spy easily without
    // having to check for specific calls.
    (executeQuery as Mock).mockClear();
    setupMovieExtraDetailsFetch();

    selectMovie(0);
    expect(executeQuery).toHaveBeenCalled();

    // after we cause something that updates the component we must call component.update to force
    // the component to render again.
    component.update();

    expect(component.find(Panel)).toHaveProp('movie', payload.data.allMovies[0]);
  });

  test('should select a movie and not fetch description and rating if available', async () => {
    const props = setup();
    const payload = setupFetchMovies(true);

    const component = shallow(
        <Home {...props} />
    );

    await flushPromises();
    const selectMovie = component.find(Movies).prop('selectMovie');

    (executeQuery as Mock).mockClear();
    selectMovie(0);
    expect(executeQuery).not.toHaveBeenCalled();

    component.update();

    expect(component.find(Panel)).toHaveProp('movie', payload.data.allMovies[0]);
  });

  test('should add a movie', async () => {
    const props = setup();
    setupFetchMovies(false);

    const component = shallow(
        <Home {...props} />
    );

    const payload: any = setupCreateMovie();

    const updateValue = component.find(Movies).prop('updateValue');
    const addMovie = component.find(Movies).prop('addMovie');

    updateValue({ target: { value: 'movie title'}});
    addMovie();

    await flushPromises();

    component.update();

    expect(component.find(Movies)).toHaveProp('data', expect.arrayContaining([payload.data.createMovie]));
  });

  test('should remove a movie', async () => {
    const props = setup();
    setupFetchMovies(false);

    const component = shallow(
        <Home {...props} />
    );

    await flushPromises();

    const removeMovie = component.find(Movies).prop('removeMovie');

    removeMovie(0);

    await flushPromises();

    component.update();

    expect(component.find(Movies)).toHaveProp('data', []);
  });

  test('should clear all movies', async () => {
    const props = setup();
    setupFetchMovies(false);

    const component = shallow(
        <Home {...props} />
    );

    await flushPromises();

    const clearAll = component.find(Movies).prop('clearAll');

    clearAll();

    await flushPromises();

    component.update();

    expect(component.find(Movies)).toHaveProp('data', []);
  });

  test('should clear the selected movie', async () => {
    const props = setup();
    const payload = setupFetchMovies(true);

    const component = shallow(
        <Home {...props} />
    );

    await flushPromises();

    const selectMovie = component.find(Movies).prop('selectMovie');
    const clearSelected = component.find(Panel).prop('clearSelected');

    (executeQuery as Mock).mockClear();
    selectMovie(0);

    component.update();

    expect(component.find(Panel)).toHaveProp('movie', payload.data.allMovies[0]);

    clearSelected();
    component.update();

    expect(component.find(Panel)).toHaveProp('movie', undefined);
  });
});
