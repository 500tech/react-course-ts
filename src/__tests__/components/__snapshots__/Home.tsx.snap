// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home/> should render 1`] = `
<div
  className="app"
>
  <Movies
    addMovie={[Function]}
    clearAll={[Function]}
    data={Array []}
    inputVal=""
    removeMovie={[Function]}
    selectMovie={[Function]}
    updateValue={[Function]}
  />
  <Panel
    clearSelected={[Function]}
    updateRating={[Function]}
  />
</div>
`;
